diff --git a/app/bundles/DashboardBundle/Event/WidgetDetailEvent.php b/app/bundles/DashboardBundle/Event/WidgetDetailEvent.php
index 3d05d425d6..1fcf7fd579 100644
--- a/app/bundles/DashboardBundle/Event/WidgetDetailEvent.php
+++ b/app/bundles/DashboardBundle/Event/WidgetDetailEvent.php
@@ -177,7 +177,9 @@ public function setTemplateData(array $templateData, $skipCache = false)
         // Store the template data to the cache
         if (!$skipCache && $this->cacheDir && $this->widget->getCacheTimeout() > 0) {
             $cache = new CacheStorageHelper($this->cacheDir, $this->uniqueCacheDir);
-            $cache->set($this->getUniqueWidgetId(), $templateData);
+            // must pass a DateTime object or a int of seconds to expire as 3rd attribute to set().
+            $expireTime = $this->widget->getCacheTimeout() * 60;
+            $cache->set($this->getUniqueWidgetId(), $templateData, (int) $expireTime);
         }
     }

