diff --git a/app/bundles/LeadBundle/Segment/Query/ContactSegmentQueryBuilder.php b/app/bundles/LeadBundle/Segment/Query/ContactSegmentQueryBuilder.php
index 3486b7df6f..3ad9520b68 100644
--- a/app/bundles/LeadBundle/Segment/Query/ContactSegmentQueryBuilder.php
+++ b/app/bundles/LeadBundle/Segment/Query/ContactSegmentQueryBuilder.php
@@ -109,8 +109,15 @@ public function assembleContactsSegmentQueryBuilder($segmentId, $segmentFilters)
      */
     public function wrapInCount(QueryBuilder $qb)
     {
+        /** @var Connection $connection */
+        $connection = $this->entityManager->getConnection();
+        if ($connection instanceof MasterSlaveConnection) {
+            // Prefer a slave connection if available.
+            $connection->connect('slave');
+        }
+
         // Add count functions to the query
-        $queryBuilder = new QueryBuilder($this->entityManager->getConnection());
+        $queryBuilder = new QueryBuilder($connection);
 
         //  If there is any right join in the query we need to select its it
         $primary = $qb->guessPrimaryLeadContactIdColumn();
