diff --git a/app/bundles/CoreBundle/Helper/CacheHelper.php b/app/bundles/CoreBundle/Helper/CacheHelper.php
index 4f07df1ba0..3681181058 100644
--- a/app/bundles/CoreBundle/Helper/CacheHelper.php
+++ b/app/bundles/CoreBundle/Helper/CacheHelper.php
@@ -173,5 +173,9 @@ protected function clearOpcaches($configSave = false)
         if (function_exists('apc_clear_cache')) {
             apc_clear_cache();
         }
+
+        if (function_exists('apcu_clear_cache')) {
+            apcu_clear_cache();
+        }
     }
 }
diff --git a/app/config/config_prod.php b/app/config/config_prod.php
index b651e891d4..04a5d1f985 100644
--- a/app/config/config_prod.php
+++ b/app/config/config_prod.php
@@ -8,21 +8,37 @@
     $loader->import('security.php');
 }
 
-/*
-$container->loadFromExtension("framework", array(
-    "validation" => array(
-        "cache" => "apc"
-    )
-));
-
-$container->loadFromExtension("doctrine", array(
-    "orm" => array(
-        "metadata_cache_driver" => "apc",
-        "result_cache_driver"   => "apc",
-        "query_cache_driver"    => "apc"
-    )
-));
-*/
+if (function_exists('apcu_store')) {
+    // Validation caching does not work in Mautic currently - https://github.com/mautic/mautic/issues/6259
+    // $container->loadFromExtension('framework', array(
+    //     'validation' => array(
+    //         'cache' => 'apcu',
+    //     )
+    // ));
+    $container->loadFromExtension('doctrine', [
+        'orm' => [
+            'metadata_cache_driver' => 'apcu',
+            'query_cache_driver'    => 'apcu',
+            // You can use APCu for result caching if using a single node in production, otherwise Redis works well.
+            'result_cache_driver'   => 'array',
+        ],
+    ]);
+} elseif (function_exists('apc_store')) {
+    // Validation caching does not work in Mautic currently - https://github.com/mautic/mautic/issues/6259
+    // $container->loadFromExtension('framework', array(
+    //     'validation' => array(
+    //         'cache' => 'apc',
+    //     )
+    // ));
+    $container->loadFromExtension('doctrine', [
+        'orm' => [
+            'metadata_cache_driver' => 'apc',
+            'query_cache_driver'    => 'apc',
+            // You can use APC for result caching if using a single node in production, otherwise Redis works well.
+            'result_cache_driver'   => 'array',
+        ],
+    ]);
+}
 
 $debugMode = $container->hasParameter('mautic.debug') ? $container->getParameter('mautic.debug') : $container->getParameter('kernel.debug');
 
diff --git a/app/config/config_test.php b/app/config/config_test.php
index dd8849ea52..ba99dca1b8 100644
--- a/app/config/config_test.php
+++ b/app/config/config_test.php
@@ -53,6 +53,24 @@
     'disable_delivery' => true,
 ]);
 
+if (function_exists('apcu_store')) {
+    $container->loadFromExtension('doctrine', [
+        'orm' => [
+            'metadata_cache_driver' => 'apcu',
+            'query_cache_driver'    => 'apcu',
+            'result_cache_driver'   => 'array',
+        ],
+    ]);
+} elseif (function_exists('apc_store')) {
+    $container->loadFromExtension('doctrine', [
+        'orm' => [
+            'metadata_cache_driver' => 'apc',
+            'query_cache_driver'    => 'apc',
+            'result_cache_driver'   => 'array',
+        ],
+    ]);
+}
+
 $container->loadFromExtension('doctrine', [
     'dbal' => [
         'default_connection' => 'default',
diff --git a/composer.json b/composer.json
index 984a81dddf..342ad326f5 100644
--- a/composer.json
+++ b/composer.json
@@ -64,7 +64,7 @@
         "doctrine/dbal": "~2.5.4",
         "doctrine/common": "~2.5.3",
         "doctrine/annotations": "1.2.*,>=1.2.7",
-        "doctrine/cache": "~1.5.4",
+        "doctrine/cache": "^1.6",
         "doctrine/migrations": "~1.2.2",
         "doctrine/orm": "~2.5.4",
         "doctrine/doctrine-bundle": "1.6.13",
diff --git a/composer.lock b/composer.lock
index 1f0abb2bb3..8b7ce35e16 100644
--- a/composer.lock
+++ b/composer.lock
@@ -322,33 +322,33 @@
         },
         {
             "name": "doctrine/cache",
-            "version": "v1.5.4",
+            "version": "v1.6.2",
             "source": {
                 "type": "git",
                 "url": "https://github.com/doctrine/cache.git",
-                "reference": "47cdc76ceb95cc591d9c79a36dc3794975b5d136"
+                "reference": "eb152c5100571c7a45470ff2a35095ab3f3b900b"
             },
             "dist": {
                 "type": "zip",
-                "url": "https://api.github.com/repos/doctrine/cache/zipball/47cdc76ceb95cc591d9c79a36dc3794975b5d136",
-                "reference": "47cdc76ceb95cc591d9c79a36dc3794975b5d136",
+                "url": "https://api.github.com/repos/doctrine/cache/zipball/eb152c5100571c7a45470ff2a35095ab3f3b900b",
+                "reference": "eb152c5100571c7a45470ff2a35095ab3f3b900b",
                 "shasum": ""
             },
             "require": {
-                "php": ">=5.3.2"
+                "php": "~5.5|~7.0"
             },
             "conflict": {
                 "doctrine/common": ">2.2,<2.4"
             },
             "require-dev": {
-                "phpunit/phpunit": ">=3.7",
+                "phpunit/phpunit": "~4.8|~5.0",
                 "predis/predis": "~1.0",
                 "satooshi/php-coveralls": "~0.6"
             },
             "type": "library",
             "extra": {
                 "branch-alias": {
-                    "dev-master": "1.5.x-dev"
+                    "dev-master": "1.6.x-dev"
                 }
             },
             "autoload": {
@@ -388,7 +388,7 @@
                 "cache",
                 "caching"
             ],
-            "time": "2015-12-19T05:03:47+00:00"
+            "time": "2017-07-22T12:49:21+00:00"
         },
         {
             "name": "doctrine/collections",
diff --git a/index_dev.php b/index_dev.php
index b4607bdd81..c1f302f41d 100644
--- a/index_dev.php
+++ b/index_dev.php
@@ -20,10 +20,15 @@
  */
 umask(0000);
 
-if (extension_loaded('apc') && in_array(@$_SERVER['REMOTE_ADDR'], ['127.0.0.1', '::1', '172.17.0.1'])) {
-    @apc_clear_cache();
-    @apc_clear_cache('user');
-    @apc_clear_cache('opcode');
+if (in_array(@$_SERVER['REMOTE_ADDR'], ['127.0.0.1', '::1', '172.17.0.1'])) {
+    if (function_exists('apc_clear_cache')) {
+        @apc_clear_cache();
+        @apc_clear_cache('user');
+        @apc_clear_cache('opcode');
+    }
+    if (function_exists('apcu_clear_cache')) {
+        @apcu_clear_cache();
+    }
 }
 
 \Mautic\CoreBundle\ErrorHandler\ErrorHandler::register('dev');
diff --git a/upgrade.php b/upgrade.php
index b66b83daa3..17db99c4bd 100644
--- a/upgrade.php
+++ b/upgrade.php
@@ -482,6 +482,10 @@ function clear_mautic_cache()
         apc_clear_cache();
     }
 
+    if (function_exists('apcu_clear_cache')) {
+        apcu_clear_cache();
+    }
+
     return true;
 }
 
